pkg/controllers/consolidation/controller.go

```diff
@@ -139,56 +136,120 @@ type candidateNode struct {
 }
 
 // ProcessCluster is exposed for unit testing purposes
-func (c *Controller) ProcessCluster(ctx context.Context) (ProcessResult, error) {
+func (c *Controller) ProcessCluster(ctx context.Context) (LifecycleResult, error) {
 	candidates, err := c.candidateNodes(ctx)
 	if err != nil {
-		return ProcessResultFailed, fmt.Errorf("determining candidate nodes, %w", err)
+		return LifecycleResultFailed, fmt.Errorf("determining candidate nodes, %w", err)
 	}
 	if len(candidates) == 0 {
-		return ProcessResultNothingToDo, nil
+		return LifecycleResultNothingToDo, nil
 	}
 
-	emptyNodes := lo.Filter(candidates, func(n candidateNode, _ int) bool { return len(n.pods) == 0 })
-	// first see if there are empty nodes that we can delete immediately, and if so delete them all at once
-	if len(emptyNodes) > 0 {
-		c.performConsolidation(ctx, consolidationAction{
-			oldNodes: lo.Map(emptyNodes, func(n candidateNode, _ int) *v1.Node { return n.Node }),
-			result:   consolidateResultDeleteEmpty,
-		})
-		return ProcessResultConsolidated, nil
-	}
+	// range over the different lifecycle methods. We'll only let one method perform an action
+	for _, fn := range []func(ctx context.Context, candidates []candidateNode) (LifecycleResult, error){
+		c.deleteEmptyNodes,
+		c.consolidateNodes,
+	} {
+		result, err := fn(ctx, candidates)
+		if err != nil {
+			return LifecycleResultFailed, err
+		}
+		switch result {
+		case LifecycleResultNothingToDo:
+			// if this method can't perform any work, try the next one
+			continue
+		case LifecyleResultRetry:
+			// The action attempted to do something, but it failed due to cluster flux. Return instead of going to
+			// lesser priority methods to it can have another attempt.
+			return LifecyleResultRetry, nil
+		case LifecycleResultFailed:
+			// The method failed in some way, it should also have returned a non-nil error and been handled above, but
+			// just in case we handle it here
+			return LifecycleResultFailed, fmt.Errorf("lifecycle generic failure")
+		case LifecycleResultSuccess:
+			// the method did something, so we stop any other methods from running and wait for the next opportunity to
+			// act
+			return LifecycleResultSuccess, nil
+		default:
+			// shouldn't occur
+			return LifecycleResultFailed, fmt.Errorf("unhandled lifecycle result %d", result)
+		}
+	}
+	return LifecycleResultNothingToDo, nil
+}
 
+// consolidateNodes looks at the non-empty nodes and executes the first viable consolidation action
+func (c *Controller) consolidateNodes(ctx context.Context, candidates []candidateNode) (LifecycleResult, error) {
```

Comment:
Discussed offline, we probably need to take a new cluster snapshot for after each simulation, especially if we're waiting for n seconds in between consolidations.